scripts/rewritepass.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/python3
+import sys
+import os
+
+if len(sys.argv) != 3:
+  print("Use: %s <opt path> <passes>" % sys.argv[0])
+  exit(1)
+
+optpath = sys.argv[1]
+passes = sys.argv[2].strip("'\"").split(',')
+plen = len(passes)
+
+level_ancestors = {
+  "module":[],
+  "cgscc":["module"],
+  "function":["module"],
+      # don't use module(cgscc(function(...)), use module(function(...))
+      # the former one results in a different output.
+      # e.g. llvm/test/Transforms/LoopVectorize/reduction-order.ll
+  "loop":["function", "module"]
+}
+
+"""
+PassBuilder.cpp states that:
+  As a special shortcut, if the very first pass is not
+  a module pass (as a module pass manager is), this will automatically form
+  the shortest stack of pass managers that allow inserting that first pass.
+Build the stack of pass managers by invoking opt with the first pass and
+parsing its error message.
+"""
+i = 0
+firstp = passes[i]
+if firstp.find("<") != -1 or firstp.find("(") != -1:
+  # make p complete by concatenating tokens until the corresponding ">"
+  # and ")" are found
+  par = par2 = 0
+  ps = []
+  while i < plen:
+    ps.append(passes[i])
+    par  += passes[i].count("(") - passes[i].count(")")
+    par2 += passes[i].count("<") - passes[i].count(">")
+    if par == 0 and par2 == 0:
+      break
+    i = i + 1
+  firstp = ",".join(ps)
+
+
+# 'opt -passes=..' does not generate the message we want if one of these passes
+# is used.
+unsupported_module_passes = [
```

Comment:
are these module passes?

Reply:
Yes, they are.
I just found that PassRegistry.def contains a useful macro (`MODULE_PASS`)

Reply:
The file is at llvm/lib. Should we update include dir at CMakeList.txt to use it?

Reply:
I see that using `opt -passes=pgo-instr-use,globaltopt` gives some weird error. But maybe just skip those that error out?

Reply:
Their messages are heterogeneous, so I'd like to simply ignore them.. 😭
I found that `insert-gcov-profiling` simply does not generate any message. I'll remove it from the list

Reply:
so if `echo "" |opt -S -passes='verify,simplify-cfg,globalopt'` doesn't produce any error, we would skip it, right?
Ok, I guess we should go and just grep the file you mention and extract the list of module passes. I hope the pass names used for the NPM are the same as for the old PM. So we can have a single skip list.

Reply:
Yes, I'll do that.
I'd like to get the cmake variable that contains the path to LLVM source, how can I find one?

Reply:
Maybe I can use  `LLVM_CMAKE_DIR` : it prints `llvm/cmake/modules`

Reply:
try LLVM_BUILD_MAIN_SRC_DIR

Reply:
Thanks! I'm debugging my code now

Reply:
For the single skip list, I'm checking it